File: scripts/openclaw-control-cli.py
#!/usr/bin/env python3
"""
openclaw-control-cli
CLI tool for OpenClaw gateway configuration and proxy management

This tool directly modifies the gateway configuration file since
control-ui requires secure context device identity.
"""

import argparse
import json
import os
import sys
from pathlib import Path
from typing import Any, Dict, Optional

# Try to load from qclaw config first
QCLAW_CONFIG = Path.home() / ".qclaw" / "openclaw.json"
OPENCLAW_CONFIG = Path.home() / ".openclaw" / "openclaw.json"

CONFIG_FILE = QCLAW_CONFIG if QCLAW_CONFIG.exists() else OPENCLAW_CONFIG


def load_config() -> Dict[str, Any]:
    """Load gateway configuration"""
    if not CONFIG_FILE.exists():
        raise FileNotFoundError(f"Config file not found: {CONFIG_FILE}")
    
    with open(CONFIG_FILE, "r", encoding="utf-8") as f:
        content = f.read()
    
    # QClaw outputs standard JSON
    return json.loads(content)


def save_config(config: Dict[str, Any], backup: bool = True):
    """Save gateway configuration"""
    if backup:
        backup_file = CONFIG_FILE.with_suffix(".json.bak")
        if CONFIG_FILE.exists():
            import shutil
            shutil.copy(CONFIG_FILE, backup_file)
    
    with open(CONFIG_FILE, "w", encoding="utf-8") as f:
        json.dump(config, f, indent=2, ensure_ascii=False)
        f.write("\n")
    
    print(f"✅ Configuration saved to {CONFIG_FILE}")
    print("ℹ️  Restart QClaw app to apply changes")


def get_nested_value(obj: dict, path: str) -> Any:
    """Get nested dictionary value"""
    keys = path.split(".")
    for key in keys:
        if isinstance(obj, dict):
            obj = obj.get(key)
        else:
            return None
    return obj


def set_nested_value(obj: dict, path: str, value: Any):
    """Set nested dictionary value"""
    keys = path.split(".")
    for key in keys[:-1]:
        obj = obj.setdefault(key, {})
    obj[keys[-1]] = value


def parse_value(value_str: str) -> Any:
    """Parse string value to appropriate type"""
    try:
        return json.loads(value_str)
    except:
        return value_str


# ──────────────────────────────────────────────────────────────────────────────
# Commands
# ──────────────────────────────────────────────────────────────────────────────

def cmd_status(args):
    """Show gateway status from config file"""
    config = load_config()
    
    print("\n📊 Gateway Configuration Summary\n")
    
    # Gateway
    gateway = config.get("gateway", {})
    print(f"  Port: {gateway.get('port', 'default')}")
    print(f"  Mode: {gateway.get('mode', 'local')}")
    print(f"  Auth: {gateway.get('auth', {}).get('mode', 'none')}")
    
    # Proxy
    proxy = gateway.get("proxy", {})
    if proxy:
        print("\n🌐 Proxy Settings\n")
        print(f"  HTTP: {proxy.get('http') or 'not set'}")
        print(f"  HTTPS: {proxy.get('https') or 'not set'}")
        print(f"  No Proxy: {proxy.get('noProxy') or 'not set'}")
    else:
        print("\n🌐 Proxy: not configured\n")
    
    # Channels
    channels = config.get("channels", {})
    if channels:
        print("\n📡 Channels\n")
        for name, cfg in channels.items():
            enabled = "✅" if cfg.get("enabled", False) else "❌"
            print(f"  {enabled} {name}")
    
    # Skills limit
    skills = config.get("skills", {})
    limits = skills.get("limits", {})
    if limits:
        print("\n📦 Skills Limits\n")
        print(f"  Max in Prompt: {limits.get('maxSkillsInPrompt', 'unlimited')}")
        print(f"  Max Prompt Chars: {limits.get('maxSkillsPromptChars', 'unlimited')}")
    
    return config


def cmd_config(args):
    """Get/set configuration values"""
    config = load_config()
    
    if not args.config_action:
        # Show full config
        print(json.dumps(config, indent=2, ensure_ascii=False))
        return config
    
    if args.config_action == "get":
        value = get_nested_value(config, args.config_path) if args.config_path else config
        print(json.dumps(value, indent=2, ensure_ascii=False) if value is not None else "null")
        return value
    
    if args.config_action == "set":
        if not args.config_path or args.config_value is None:
            print("Error: --path and --value required for set")
            sys.exit(1)
        
        value = parse_value(args.config_value)
        set_nested_value(config, args.config_path, value)
        save_config(config)
        
        print(f"✅ Set {args.config_path} = {json.dumps(value)}")
        return config


def cmd_proxy(args):
    """Manage proxy settings"""
    config = load_config()
    
    # Ensure gateway.proxy exists
    config.setdefault("gateway", {}).setdefault("proxy", {})
    proxy = config["gateway"]["proxy"]
    
    if not args.proxy_action or args.proxy_action == "show":
        print("\n🌐 Proxy Configuration\n")
        print(f"  HTTP: {proxy.get('http') or 'not set'}")
        print(f"  HTTPS: {proxy.get('https') or 'not set'}")
        print(f"  No Proxy: {proxy.get('noProxy') or 'not set'}")
        return proxy
    
    if args.proxy_action == "set":
        if not args.proxy_url:
            print("Error: Proxy URL required")
            sys.exit(1)
        
        proxy_url = args.proxy_url
        
        # Parse proxy URL
        from urllib.parse import urlparse
        parsed = urlparse(proxy_url)
        
        # Set HTTP and HTTPS proxy
        proxy["http"] = proxy_url
        proxy["https"] = proxy_url
        
        # Handle no_proxy from query params
        if parsed.query:
            import urllib.parse as up
            params = dict(up.parse_qsl(parsed.query))
            if "noProxy" in params:
                proxy["noProxy"] = params["noProxy"]
        
        save_config(config)
        print(f"✅ Proxy configured: {proxy_url}")
        return proxy
    
    if args.proxy_action == "clear":
        if "proxy" in config.get("gateway", {}):
            del config["gateway"]["proxy"]
            save_config(config)
            print("✅ Proxy cleared")
        else:
            print("ℹ️  No proxy configured")
        
        return {"cleared": True}


def cmd_edit(args):
    """Open config file in editor"""
    editor = os.environ.get("EDITOR", "nano")
    os.system(f"{editor} {CONFIG_FILE}")


# ──────────────────────────────────────────────────────────────────────────────
# Main
# ──────────────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(
        description="OpenClaw Control CLI - Gateway configuration and proxy management",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=f"""
Config File: {CONFIG_FILE}

Examples:
  # Show configuration summary
  %(prog)s status

  # Show proxy settings
  %(prog)s proxy show

  # Set proxy
  %(prog)s proxy set http://127.0.0.1:7890

  # Set proxy with no_proxy
  %(prog)s proxy set "http://127.0.0.1:7890?noProxy=localhost,127.0.0.1,.local"

  # Clear proxy
  %(prog)s proxy clear

  # Get config value
  %(prog)s config get gateway.port

  # Set config value
  %(prog)s config set gateway.port 28790

  # Open config in editor
  %(prog)s edit
""")
    
    subparsers = parser.add_subparsers(dest="command", help="Available commands")
    
    # Status
    subparsers.add_parser("status", help="Show gateway configuration summary")
    
    # Config
    p_config = subparsers.add_parser("config", help="Get/set configuration")
    p_config.add_argument("config_action", choices=["get", "set"], nargs="?", help="Action")
    p_config.add_argument("config_path", nargs="?", help="Config path (dot notation)")
    p_config.add_argument("config_value", nargs="?", help="Value to set (JSON or string)")
    
    # Proxy
    p_proxy = subparsers.add_parser("proxy", help="Manage proxy settings")
    p_proxy.add_argument("proxy_action", choices=["show", "set", "clear"], nargs="?", help="Action")
    p_proxy.add_argument("proxy_url", nargs="?", help="Proxy URL")
    
    # Edit
    subparsers.add_parser("edit", help="Open config file in editor")
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        sys.exit(0)
    
    try:
        if args.command == "status":
            cmd_status(args)
        elif args.command == "config":
            cmd_config(args)
        elif args.command == "proxy":
            cmd_proxy(args)
        elif args.command == "edit":
            cmd_edit(args)
    except FileNotFoundError as e:
        print(f"Error: {e}")
        print(f"Make sure QClaw is running and {CONFIG_FILE} exists")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/openclaw-control-cli.js
#!/usr/local/opt/node@24/bin/node
/**
 * openclaw-control-cli
 * CLI tool for OpenClaw gateway control and proxy connection
 * 
 * Usage:
 *   node openclaw-control-cli.js <command> [options]
 * 
 * Commands:
 *   connect      Connect to gateway with optional proxy
 *   status       Show gateway status
 *   config       Get/set gateway configuration
 *   proxy        Configure proxy settings
 *   channels     List channel status
 *   sessions     List active sessions
 */

import { WebSocket } from 'ws';

const GATEWAY_URL = process.env.OPENCLAW_GATEWAY_URL || 'ws://127.0.0.1:28789';
const DEFAULT_TIMEOUT = 10000;

// ──────────────────────────────────────────────────────────────────────────────
// Gateway Client
// ──────────────────────────────────────────────────────────────────────────────

class GatewayClient {
  constructor(url, options = {}) {
    this.url = url;
    this.token = options.token || process.env.OPENCLAW_TOKEN || '';
    this.password = options.password || process.env.OPENCLAW_PASSWORD || '';
    this.timeout = options.timeout || DEFAULT_TIMEOUT;
    this.ws = null;
    this.pending = new Map();
    this.seq = 0;
    this.connected = false;
    this.hello = null;
  }

  connect() {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => reject(new Error('Connection timeout')), this.timeout);
      
      this.ws = new WebSocket(this.url);
      
      this.ws.on('open', () => {
        clearTimeout(timer);
        this.sendConnect();
      });
      
      this.ws.on('message', (data) => {
        try {
          const msg = JSON.parse(String(data));
          this.handleMessage(msg, resolve, reject);
        } catch (e) {
          console.error('Parse error:', e.message);
        }
      });
      
      this.ws.on('error', (err) => {
        clearTimeout(timer);
        reject(new Error(`WebSocket error: ${err.message}`));
      });
      
      this.ws.on('close', (code, reason) => {
        this.connected = false;
        if (!this.hello) {
          reject(new Error(`Connection closed: ${code} ${reason}`));
        }
      });
    });
  }

  sendConnect() {
    const payload = {
      minProtocol: 3,
      maxProtocol: 3,
      client: {
        id: 'openclaw-control-cli',
        version: '1.0.0',
        platform: process.platform,
        mode: 'cli',
      },
      role: 'operator',
      scopes: ['operator.admin', 'operator.approvals', 'operator.pairing'],
      auth: this.token ? { token: this.token } : this.password ? { password: this.password } : undefined,
      userAgent: 'openclaw-control-cli/1.0.0',
      locale: process.env.LANG || 'en-US',
    };
    
    this.request('connect', payload);
  }

  handleMessage(msg, resolve, reject) {
    if (msg.type === 'res') {
      const pending = this.pending.get(msg.id);
      if (pending) {
        this.pending.delete(msg.id);
        if (msg.ok) {
          pending.resolve(msg.payload);
        } else {
          pending.reject(new Error(msg.error?.message || 'Request failed'));
        }
      }
      return;
    }
    
    if (msg.type === 'event') {
      if (msg.event === 'connect.hello') {
        this.hello = msg.payload;
        this.connected = true;
        resolve(msg.payload);
      }
      // Handle other events as needed
      return;
    }
  }

  request(method, params = {}) {
    return new Promise((resolve, reject) => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        reject(new Error('Not connected'));
        return;
      }
      
      const id = `cli-${++this.seq}`;
      const msg = { type: 'req', id, method, params };
      
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`Request timeout: ${method}`));
      }, this.timeout);
      
      this.pending.set(id, {
        resolve: (payload) => {
          clearTimeout(timer);
          resolve(payload);
        },
        reject: (err) => {
          clearTimeout(timer);
          reject(err);
        },
      });
      
      this.ws.send(JSON.stringify(msg));
    });
  }

  close() {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

// ──────────────────────────────────────────────────────────────────────────────
// Commands
// ──────────────────────────────────────────────────────────────────────────────

async function cmdStatus(client) {
  const health = await client.request('health', {});
  const sessions = await client.request('sessions.list', {});
  
  console.log('\n📊 Gateway Status\n');
  console.log('  Status:', health.ok ? '✅ OK' : '❌ Error');
  console.log('  Version:', health.serverVersion || 'unknown');
  console.log('  Uptime:', health.uptime ? formatUptime(health.uptime) : 'unknown');
  console.log('  Active Sessions:', sessions.sessions?.length || 0);
  
  if (sessions.sessions?.length > 0) {
    console.log('\n  Recent Sessions:');
    sessions.sessions.slice(0, 5).forEach(s => {
      console.log(`    - ${s.key} (${s.agentId || 'main'})`);
    });
  }
  
  return { health, sessions };
}

async function cmdConfig(client, args) {
  if (args.length === 0) {
    const config = await client.request('config.get', {});
    console.log('\n⚙️  Current Configuration\n');
    console.log(JSON.stringify(config.config, null, 2));
    return config;
  }
  
  const [action, path, value] = args;
  
  if (action === 'get') {
    const result = await client.request('config.get', {});
    const value = path?.split('.').reduce((obj, key) => obj?.[key], result.config);
    console.log(value !== undefined ? JSON.stringify(value, null, 2) : 'Path not found');
    return result;
  }
  
  if (action === 'set' && path && value !== undefined) {
    const config = await client.request('config.get', {});
    const parsedValue = tryParseValue(value);
    setNestedValue(config.config, path.split('.'), parsedValue);
    
    const result = await client.request('config.set', {
      raw: JSON.stringify(config.config, null, 2),
      baseHash: config.hash,
    });
    
    console.log('✅ Configuration updated');
    return result;
  }
  
  console.log('Usage: config [get <path> | set <path> <value>]');
}

async function cmdProxy(client, args) {
  const [action, proxyUrl] = args;
  
  if (action === 'show' || !action) {
    const config = await client.request('config.get', {});
    const proxy = config.config?.gateway?.proxy;
    
    console.log('\n🌐 Proxy Configuration\n');
    console.log('  HTTP Proxy:', proxy?.http || 'not set');
    console.log('  HTTPS Proxy:', proxy?.https || 'not set');
    console.log('  No Proxy:', proxy?.noProxy || 'not set');
    
    return { proxy };
  }
  
  if (action === 'set' && proxyUrl) {
    const config = await client.request('config.get', {});
    
    config.config = config.config || {};
    config.config.gateway = config.config.gateway || {};
    config.config.gateway.proxy = config.config.gateway.proxy || {};
    
    // Parse proxy URL
    const parsed = parseProxyUrl(proxyUrl);
    Object.assign(config.config.gateway.proxy, parsed);
    
    const result = await client.request('config.set', {
      raw: JSON.stringify(config.config, null, 2),
      baseHash: config.hash,
    });
    
    console.log('✅ Proxy configured:', proxyUrl);
    return result;
  }
  
  if (action === 'clear') {
    const config = await client.request('config.get', {});
    
    if (config.config?.gateway?.proxy) {
      delete config.config.gateway.proxy;
      
      const result = await client.request('config.set', {
        raw: JSON.stringify(config.config, null, 2),
        baseHash: config.hash,
      });
      
      console.log('✅ Proxy cleared');
      return result;
    }
    
    console.log('ℹ️  No proxy configured');
    return { cleared: false };
  }
  
  console.log('Usage: proxy [show | set <url> | clear]');
}

async function cmdChannels(client) {
  const result = await client.request('channels.status', { probe: false, timeoutMs: 5000 });
  
  console.log('\n📡 Channel Status\n');
  
  const channels = result.channels || {};
  for (const [name, status] of Object.entries(channels)) {
    const icon = status.enabled ? '✅' : '❌';
    console.log(`  ${icon} ${name}: ${status.status || 'unknown'}`);
  }
  
  return result;
}

async function cmdSessions(client) {
  const result = await client.request('sessions.list', {});
  
  console.log('\n💬 Active Sessions\n');
  
  const sessions = result.sessions || [];
  if (sessions.length === 0) {
    console.log('  No active sessions');
    return result;
  }
  
  sessions.forEach((s, i) => {
    const agent = s.agentId || 'main';
    const messages = s.messageCount || 0;
    console.log(`  ${i + 1}. ${s.key} [${agent}] (${messages} messages)`);
  });
  
  return result;
}

async function cmdConnect(client, args) {
  const [proxyUrl] = args;
  
  console.log('\n🔗 Connecting to Gateway...\n');
  console.log('  URL:', client.url);
  console.log('  Auth:', client.token ? 'token' : client.password ? 'password' : 'none');
  
  try {
    const hello = await client.connect();
    console.log('\n✅ Connected!\n');
    console.log('  Server:', hello.snapshot?.serverVersion || 'unknown');
    console.log('  Sessions:', hello.snapshot?.sessions?.count || 0);
    
    if (proxyUrl) {
      console.log('\n🌐 Configuring proxy...');
      await cmdProxy(client, ['set', proxyUrl]);
    }
    
    return hello;
  } catch (err) {
    console.error('\n❌ Connection failed:', err.message);
    throw err;
  }
}

// ──────────────────────────────────────────────────────────────────────────────
// Utilities
// ──────────────────────────────────────────────────────────────────────────────

function formatUptime(seconds) {
  const d = Math.floor(seconds / 86400);
  const h = Math.floor((seconds % 86400) / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = seconds % 60;
  
  if (d > 0) return `${d}d ${h}h ${m}m`;
  if (h > 0) return `${h}h ${m}m ${s}s`;
  if (m > 0) return `${m}m ${s}s`;
  return `${s}s`;
}

function tryParseValue(str) {
  try {
    return JSON.parse(str);
  } catch {
    return str;
  }
}

function setNestedValue(obj, path, value) {
  const last = path.pop();
  const target = path.reduce((obj, key) => {
    obj[key] = obj[key] || {};
    return obj[key];
  }, obj);
  target[last] = value;
}

function parseProxyUrl(url) {
  const parsed = new URL(url);
  const result = {};
  
  if (parsed.protocol === 'http:') {
    result.http = url;
    result.https = url; // Use same for HTTPS by default
  } else if (parsed.protocol === 'https:') {
    result.https = url;
  } else if (parsed.protocol === 'socks5:') {
    result.http = url;
    result.https = url;
  }
  
  // Handle no_proxy from URL query params
  const noProxy = parsed.searchParams.get('noProxy');
  if (noProxy) {
    result.noProxy = noProxy;
  }
  
  return result;
}

// ──────────────────────────────────────────────────────────────────────────────
// CLI Entry Point
// ──────────────────────────────────────────────────────────────────────────────

async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0 || args[0] === '--help' || args[0] === '-h') {
    console.log(`
OpenClaw Control CLI

Usage:
  node openclaw-control-cli.js <command> [options]

Commands:
  connect [proxy_url]  Connect to gateway, optionally configure proxy
  status               Show gateway status
  config [get|set]     Get/set gateway configuration
  proxy [show|set|clear]  Manage proxy settings
  channels             List channel status
  sessions             List active sessions

Environment Variables:
  OPENCLAW_GATEWAY_URL   Gateway WebSocket URL (default: ws://127.0.0.1:28789)
  OPENCLAW_TOKEN         Authentication token
  OPENCLAW_PASSWORD     Authentication password

Examples:
  # Connect and show status
  node openclaw-control-cli.js connect

  # Connect with proxy
  node openclaw-control-cli.js connect http://127.0.0.1:7890

  # Show proxy settings
  node openclaw-control-cli.js proxy show

  # Set proxy
  node openclaw-control-cli.js proxy set http://127.0.0.1:7890

  # Clear proxy
  node openclaw-control-cli.js proxy clear

  # Get config value
  node openclaw-control-cli.js config get gateway.port

  # Set config value
  node openclaw-control-cli.js config set gateway.port 28790
`);
    process.exit(0);
  }
  
  const [command, ...cmdArgs] = args;
  const client = new GatewayClient(GATEWAY_URL);
  
  try {
    // Commands that don't require connection
    if (command === 'connect') {
      await cmdConnect(client, cmdArgs);
    } else {
      // All other commands need connection
      await client.connect();
      
      switch (command) {
        case 'status':
          await cmdStatus(client);
          break;
        case 'config':
          await cmdConfig(client, cmdArgs);
          break;
        case 'proxy':
          await cmdProxy(client, cmdArgs);
          break;
        case 'channels':
          await cmdChannels(client);
          break;
        case 'sessions':
          await cmdSessions(client);
          break;
        default:
          console.error(`Unknown command: ${command}`);
          console.log('Run with --help for usage');
          process.exit(1);
      }
    }
  } catch (err) {
    console.error('\n❌ Error:', err.message);
    process.exit(1);
  } finally {
    client.close();
  }
}

main().catch(console.error);
